test/object-store/sync/session/connection_change_notifications.cpp

```diff
@@ -64,20 +72,13 @@ TEST_CASE("sync: Connection state changes", "[sync][session][connection change]"
             return sessions_are_connected(*session);
         });
 
-        std::atomic<bool> listener1_called(false);
-        std::atomic<bool> listener2_called(false);
-        auto token1 = session->register_connection_change_callback(
-            [&](SyncSession::ConnectionState, SyncSession::ConnectionState) {
-                listener1_called = true;
-            });
-        session->unregister_connection_change_callback(token1);
         session->register_connection_change_callback([&](SyncSession::ConnectionState, SyncSession::ConnectionState) {
             listener2_called = true;
         });
 
         user->log_out();
         REQUIRE(sessions_are_disconnected(*session));
-        REQUIRE(listener1_called == false);
+        REQUIRE(listener1_call_cnt == 1);
```

Comment:
The way I read this test is that it is specifically supposed to check that an unregistered connection change callback is never called. But this changes the test to not test that anymore, am I missing something?

Reply:
There are actually two cases that should be tested here and maybe this should be another SECTION - there is the case where the callback is unregistered when the callback is called (was not explicitly tested before); and then there is the case of registering and then unregistering a callback and (hopefully) a connection change event has not occurred before the callback can be unregistered.

There is a tiny window where the callback could grab the mutex and unregister the callback while the callback is being executed. I'll update the previous test and add a second SECTION to test the unregister during callback.